### PR TITLE
Feature debug

### PR DESCRIPTION
**Pull request by Pandabear1125**
# This adds quite a bit
 - Adds Dual Serial support, along side normal RawHID support
 - Adds TeensyDebug library
   - Includes support for gdb and in-code debugging
 - Updates Makefile to be more readable
 - Updated the .gitignore to include what it should've included a year ago (whoops)
   - Removes the library files from being included
 - Added scripts to automate monitoring and gdb starting

### CODE REVIEW

**Comment by Pandabear1125** [comment]
Need to test on robot, this introduces very critical changes

**Comment by Pandabear1125** [comment]
> Need to test on robot, this introduces very critical changes

Tested, hid comms remain functional, there is a patch required for Hive to recognize Teensy's HID signature though.

File: teensy4/usb_desc.h
Comment (Rishi-H) [inline teensy4/usb_desc.h:1]: why does rawhid transfer and receive go down to 64?

File: tools/monitor.sh
Comment (Rishi-H) [inline tools/monitor.sh:1]: Did we not have this before? What runs on tycmd monitor? or are you just making a firmware tools dir so we don't have to use tycmd?

Comment (Pandabear1125) [reply tools/monitor.sh:1]: We did have this, however since we are switching to true Serial, tycmd monitor kinda freaks out, especially since we are now running 2 Serial lines. One line is entirely for GDB support (SerialUSB1), and the other is normal Serial (Serial). This program is needed to differentiate between the two and automatically monitor the normal Serial line, rather than SerialUSB1.

File: tools/monitor.sh
Comment (StudentKyle) [inline tools/monitor.sh:5]: don't we want it to break when you ctrl-C? This seems like we should avoid ever doing this.

Comment (StudentKyle) [reply tools/monitor.sh:5]: If it's causing some horrible behavior we definitely don't want, I think there could be a way to fix that over just saying "no canceling allowed"

File: teensy4/usb_desc.h
Comment (StudentKyle) [reply teensy4/usb_desc.h:1]: If it's unused it would be better not including it at all so people don't get confused.

File: src/main.cpp
Comment (StudentKyle) [inline src/main.cpp:87]: strange indentation. Is that intended?

File: tools/monitor.sh
Comment (StudentKyle) [reply tools/monitor.sh:5]: also have to make sure we're not leaving a zombie monitor process by doing ctrl D instead.

**Review by StudentKyle — CHANGES_REQUESTED** [review]
duplicate code between this repo and hive, need to choose one repo and keep it there so we don't have to maintain both.

Also see comment about trapping ctrl-c

File: tools/monitor.sh
Comment (StudentKyle) [reply tools/monitor.sh:5]: You said you verified that there's no zombie processes, so that's good.

Still want to check this myself, and see if I can make ctrl-c work properly as soon as I get the chance. If this is the only thing holding up the approval let me know.

File: tools/prepare_gdb.sh
Comment (StudentKyle) [inline tools/prepare_gdb.sh:1]: duplicate code between `monitor.sh` and `prepare_gdb.sh`.

To make it non-duplicate you could have a script called `get_tty_path.sh` and have it stdout the result. Then process them separately as you do at the end, with:

```bash
tty_path=$(./tools/get_tty_path.sh)

... # same code as before to process it
```
( Errors can be captured as well, but probably not needed: [capture stderr](https://stackoverflow.com/questions/12738460/how-to-get-output-of-a-bash-command-in-a-variable) )

I recommend this just to minimize maintainability issues.

File: src/main.cpp
Comment (StudentKyle) [inline src/main.cpp:17]: stray include in main, I don't see it being used here.

**Review by StudentKyle — CHANGES_REQUESTED** [review]
Some maintanability comments and other minor oversights

File: src/main.cpp
Comment (Pandabear1125) [reply src/main.cpp:17]: The "debug" object comes from this

File: .gitignore
Comment (Pandabear1125) [reply .gitignore:16]: git scraper and gdb preparer

File: .gitignore
Comment (StudentKyle) [reply .gitignore:3]: maybe not a problem if vscode extension doesn't automatically activate without cmake.

File: Makefile
Comment (Pandabear1125) [reply Makefile:162]: That sounds good.

Yes, its a prereq target for build.

File: src/main.cpp
Comment (Pandabear1125) [reply src/main.cpp]: me when tabs

File: Makefile
Comment (StudentKyle) [inline Makefile:164]: me when spacing

File: tools/prepare_gdb.sh
Comment (StudentKyle) [inline tools/prepare_gdb.sh:1]: (bump) 

duplicate code between `monitor.sh` and `prepare_gdb.sh`.

To make it non-duplicate you could have a script called `get_tty_path.sh` and have it stdout the result. Then process them separately as you do at the end, with:

```bash
tty_path=$(./tools/get_tty_path.sh)

... # same code as before to process it
```
( Errors can be captured as well, but probably not needed: [capture stderr](https://stackoverflow.com/questions/12738460/how-to-get-output-of-a-bash-command-in-a-variable) )

I recommend this just to minimize maintainability issues.

**Review by StudentKyle — CHANGES_REQUESTED** [review]
me when spacing

**Comment by StudentKyle** [comment]
Also looks like there's a `firmware.hex` file. Likely a remnant from before it was added to .gitignore

File: tools/prepare_gdb.sh
Comment (StudentKyle) [reply tools/prepare_gdb.sh:1]: @Pandabear1125 
Took another look at this. Here's my implementation (tested functionality of this script, but not with teensy):
`tools/get_tty_path.sh`
``` bash
#!/bin/bash

# This finds the correct ttyACM device. First argument passed in should be the device ID.
# Example usage: tty_path=$(./tools/get_tty_path.sh *-f00)

# path to the /dev/ serial device that the Teensy is connected to
tty_path=""

if [ ! -n "$1" ]; then
    # 1>&2 redirects stdout (1) into stderr (2)
    echo "\"$1\" is not a valid tty ID" 1>&2
    exit 1
fi

# Find the correct serial device
for i in /dev/serial/by-id/*; do
    # compare with first argument passed to this command
	if [[ "$i" == $1 ]]; then
		tty_path=$i
		break
	fi
done

# found the id, now we need the actual path to the specific ttyACM device
if [ -n "$tty_path" ]; then
	tty_path=$(readlink -f $tty_path)
fi

# if not valid
if [ ! -n "$tty_path" ]; then
    # 1>&2 redirects stdout (1) into stderr (2)
    echo "Failed to detect tty_path with id \"$1\"" 1>&2
    echo "This probably means the Teensy is not connected or the Teensy is in bootloader mode (red LED on/blinking)" 1>&2

    exit 1
fi

echo $tty_path # outputs to stdout by default

exit 0
```

Comment (StudentKyle) [reply tools/prepare_gdb.sh:1]: Now that I see it it's OK, but might make some of the existing comments more confusing in the gdb and monitor files when describing the ID. 

So up to you if you want this, but I wanted to give it a shot just to learn (will do the nmcli ethernet fix next, this was practice)

Comment (Pandabear1125) [reply tools/prepare_gdb.sh:1]: Could you explain what you find confusing in the gdb and monitor files regarding the ID?

**Review by StudentKyle — APPROVED** [review]
See comment https://github.com/CU-Robotics/firmware/pull/65#discussion_r1744330393

Adding that is up to you like I said, just wanted to give it a try. Since it's only that and the ctrl-c issue I'll approve this. 

If I get time later in the year I might revisit the ctrl-c issue, but your warning about it at the start of monitoring seems sufficient (as long as you have time to read it?). We should make a github issue for it.

Still need to test the debugging and new scripts on a robot/teensy once someone gets a chance, so **do not merge** until someone else or me tests it.

File: tools/prepare_gdb.sh
Comment (StudentKyle) [reply tools/prepare_gdb.sh:1]: I think the comments are very good. I meant it might be slightly less clear if it reads:

`prepare_gdb.sh`
```bash
# The correct serial device has the following name
# usb-Teensyduino_USB_Custom_*-if02
# Where * is a number unique to that serial device, we dont care about that though
# The important part is the "if02" at the end, its interface number 02
# SerialUSB1 is what interface 02 references

tty_path=$(./tools/get_tty_path.sh *-if02)
```

`monitor.sh`
```bash
# The correct serial device has the following name
# usb-Teensyduino_USB_Custom_*-if00
# where * is a number
# The important part is the "if00" at the end, its interface number 00

tty_path=$(./tools/get_tty_path.sh *-if00)
```

Actually maybe that is better. Maybe would add a comment saying (in the get_tty_path.sh I put above)

```bash
# Find the correct serial device.
# The important part is the ID/interface number at the end, which is compared here
for i in /dev/serial/by-id/*; do
    ...
fi
```

Comment (StudentKyle) [reply tools/prepare_gdb.sh:1]: Also one note is I'm not sure if the script comparison should be 

```bash
[[ "$i" == $1 ]]
```
or
```bash
[[ "$i" == "$1" ]]
```

Couldn't really test it without a teensy, so check that syntax.

Comment (Pandabear1125) [reply tools/prepare_gdb.sh:1]: I tested it, it works. Feel free to add it

**Comment by Pandabear1125** [comment]
Tested on infantry, merging

**Pull request merged**
